src/iocost/iocost.c

```diff
@@ -0,0 +1,298 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <errno.h>
+#include <getopt.h>
+#include <stdio.h>
+#include <unistd.h>
+
+#include "sd-device.h"
+#include "sd-hwdb.h"
+
+#include "alloc-util.h"
+#include "cgroup-util.h"
+#include "devnum-util.h"
+#include "main-func.h"
+#include "path-util.h"
+#include "pretty-print.h"
+#include "verbs.h"
+
+#define DEFAULT_SOLUTION "isolatedbandwidth"
+
+static int help(void) {
+        _cleanup_free_ char *link = NULL;
+        int r;
+
+        r = terminal_urlify_man("systemd-iocost", "1", &link);
+        if (r < 0)
+                return log_oom();
+
+        printf("%s [OPTIONS...]\n\n"
+               "Set up iocost model and qos solutions for block devices from hwdb\n"
+               "\nCommands:\n"
+               "  apply <path> [solution]    Apply the specified solution to the device represented by <path>\n"
+               "                             (solution defaults to "DEFAULT_SOLUTION")\n"
+               "  query <path>               Query hwdb and print known solutions for the device represented\n"
+               "                             by <path>\n"
+               "\nOptions:\n"
+               "  -h --help                  Show this help\n"
+               "     --version               Show package version\n"
+               "\nSee the %s for details.\n",
+               program_invocation_short_name,
+               link);
+
+        return 0;
+}
+
+static int parse_argv(int argc, char *argv[]) {
+
+        enum {
+                ARG_VERSION = 0x100,
+        };
+
+        static const struct option options[] = {
+                { "help",      no_argument,       NULL, 'h'           },
+                { "version",   no_argument,       NULL, ARG_VERSION   },
+                {}
+        };
+
+        int c;
+
+        assert(argc >= 1);
+        assert(argv);
+
+        while ((c = getopt_long(argc, argv, "h", options, NULL)) >= 0)
+                switch (c) {
+
+                case 'h':
+                        return help();
+
+                case ARG_VERSION:
+                        return version();
+
+                case '?':
+                        return -EINVAL;
+
+                default:
+                        assert_not_reached();
+                }
+
+        return 1;
+}
+
+enum HwdbParseState {
+        MODEL,
+        QOS,
+        SKIP,
+        DONE,
+};
```

Comment:
Oh, I passed over the example in the PR description somehow.

> `block:*:*:name:PC611 NVMe SK hynix 1TB`

This needs to end in `:`, so that it is possible to extend the modalias with other fields in the future, see other files under `hwdb.d/`. (Here, we are providing a match *pattern*, and the modalias is a string we match against the pattern. Modalias strings should also end in `:` so it is possible to extend the modalias strings and the patterns in the future.)

> `IOCOST_TUNE_BANDWIDTH_MODEL=rbps=891487860 rrandiops=57728 rseqiops=57906 wbps=360510501 wrandiops=9139 wseqiops=5071`
> `IOCOST_TUNE_BANDWIDTH_QOS=max=100.0 min=100.0 rlat=914 rpct=0.0 wlat=6782 wpct=0.0`
> `IOCOST_TUNE_ISOLATEDBANDWIDTH_MODEL=rbps=891487860 rrandiops=57728 rseqiops=57906 `wbps=360510501 wrandiops=9139 wseqiops=5071`
> `IOCOST_TUNE_ISOLATEDBANDWIDTH_QOS=max=100.0 min=100.0 rlat=914 rpct=0.0 wlat=6782 wpct=0.0`

So… the hwdb contains multiple "models", and we could select one. This is OK for development, but I think that for the user we just want to pick one model that works best.

I'd propose the following layout:
```
block:*:*:model:PC611 NVMe SK hynix 1TB:
  IOCOST_MODEL=rbps=891487860 rrandiops=57728 rseqiops=57906 wbps=360510501 wrandiops=9139 wseqiops=5071
  IOCOST_QOS=min=100.0 max=100.0 rlat=914 rpct=0.0 wlat=6782 wpct=0.0
```

Both properties must be optional, and any of the fields must be optional too, and suitable defaults should be assumed or calculated if some data is missing. This way we can reasonably extend this in the future and don't need to fill it in with bogus data if we don't know some value.

(Another issue to consider is that with the current layout it is not possible to override individual fields. For example, if `PC611 NVMe SK hynix 2TB` appears on the market and it is very similar to `…1TB` except that `rbps` is much higher, we can only configure the full `IOCOST_MODEL` line. But I think that in practice you'd do that anyway, so I think it's fine to just keep everything in one line for brevity.)

And when matching, just query the property, parse it, and apply. The approach with SD_HWDB_FOREACH_PROPERTY is not good.

Reply:
Hey, so, the QoS params describe a fairly wide range of tradeoffs that users may reasonably want depending on usecases. For example, by default, we'll likely want to preserve most of bandwidth while trying to not lose control of the device completely. However, on most mainstream devices, that throttling point can be way off from the point where workloads can be isolated well - e.g. the point where an interactive desktop user or latency sensitive workloads won't notice much no matter how hammered the SSD is by something else on the system. Depending on the usecase and preference, both choices are completely reasonable, so I think we need to allow the users to pick and switch the QoS target.

As for overriding the parameters partially for the same model of device with different capacities, I'm afraid most devices behave quite different depending on the capacity. Higher capacity drives usually have higher levels of parallelism, more room for GC and all the other tricks that the controllers play and tend to perform better, and SSDs usually are too complex black boxes to selective alter individual parameters. We just have to run them through the full suite of benchmarks unfortunately.

Reply:
> Something I have discussed with @htejun is maybe we can move the choice related to use case to hwdb download time, so we would generate different hwdb files per use case, and we can have the database download/update tool make that selection.

Please no ;) The whole of the hwdb is that it's a relatively compact db that you can install everywhere. I really hope we can make those config files something that we can distribute widely and get roughly-ok behaviour out-of-the-box on most machines. Generating files on the the fly would create complexity and I'd only make sense in a datacenter.

Reply:
After some more discussion, including between @htejun and @poettering elsewhere, I'm told, I am implementing a solution that covers all of the bases, hopefully. It ships all of the solutions, so no use case left behind, in the same hwdb file, so no file generation or use-case-specific download tool, and it allows specification of what the sane default by listing the solutions in order of preference. The hwdb file looks like this:

```
block:*:name:Fedora_Linux_34-0_SSD:
  IOCOST_SOLUTIONS=isolation isolated-bandwidth bandwidth naive
  IOCOST_MODEL_ISOLATION=rbps=626708081 rseqiops=27119 rrandiops=26638 wbps=138402498 wseqiops=3257 wrandiops=3208
  IOCOST_QOS_ISOLATION=rpct=0.00 rlat=1997 wpct=0.00 wlat=50837 min=100.00 max=100.00
  IOCOST_MODEL_ISOLATED-BANDWIDTH=rbps=997831946 rseqiops=43179 rrandiops=42412 wbps=220361662 wseqiops=5186 wrandiops=5107
  IOCOST_QOS_ISOLATED-BANDWIDTH=rpct=0.00 rlat=2213 wpct=0.00 wlat=61487 min=100.00 max=100.00
  IOCOST_MODEL_BANDWIDTH=rbps=1891663391 rseqiops=81857 rrandiops=80403 wbps=417755805 wseqiops=9832 wrandiops=9682
  IOCOST_QOS_BANDWIDTH=rpct=0.00 rlat=7034 wpct=0.00 wlat=87136 min=100.00 max=100.00
  IOCOST_MODEL_NAIVE=rbps=1891663391 rseqiops=81857 rrandiops=80403 wbps=417755805 wseqiops=9832 wrandiops=9682
  IOCOST_QOS_NAIVE=rpct=99.00 rlat=7034 wpct=99.00 wlat=87136 min=75.00 max=100.00
```

So what systemd-iocost does is, unless it is given a specific solution to apply, first look at IOCOST_SOLUTIONS and pick the first one, then it gets the properties for that particular solution (no FOREACH involved) and applies that. The apply verb can take an optional solution name.

The query verb is readded and now shows the known solutions for that device:

```
kov@uva ...>  ./build/systemd-iocost query /dev/sda
Known solutions for /dev/sda

[isolation]
io.cost.model: rbps=626708081 rseqiops=27119 rrandiops=26638 wbps=138402498 wseqiops=3257 wrandiops=3208
io.cost.qos: rpct=0.00 rlat=1997 wpct=0.00 wlat=50837 min=100.00 max=100.00

[isolated-bandwidth]
io.cost.model: rbps=997831946 rseqiops=43179 rrandiops=42412 wbps=220361662 wseqiops=5186 wrandiops=5107
io.cost.qos: rpct=0.00 rlat=2213 wpct=0.00 wlat=61487 min=100.00 max=100.00

[bandwidth]
io.cost.model: rbps=1891663391 rseqiops=81857 rrandiops=80403 wbps=417755805 wseqiops=9832 wrandiops=9682
io.cost.qos: rpct=0.00 rlat=7034 wpct=0.00 wlat=87136 min=100.00 max=100.00

[naive]
io.cost.model: rbps=1891663391 rseqiops=81857 rrandiops=80403 wbps=417755805 wseqiops=9832 wrandiops=9682
io.cost.qos: rpct=99.00 rlat=7034 wpct=99.00 wlat=87136 min=75.00 max=100.00
```
Thoughts? I'll push an updated, squashed branch later today or tomorrow.